### PR TITLE
Incremental compilation PoC

### PR DESCRIPTION
**Pull request by paulbutcher**
This is definitely not ready to be merged! 

But it is a proof of concept for the quick-and-dirty incremental compilation approach discussed on Gitter recently. The impact is minimal (just a few very minor changes to the language server, no changes to the compiler whatsoever).

And it works:

Screenshot 2022-06-14 at 08 55 27

### CODE REVIEW

**Comment by paulbutcher** [comment]
I completely understand where that skepticism comes from (I'm under no illusions that this is a true fix). But I do think that it's an overall win. Consider:

1. The changes are purely within the language server, and very well contained, so can be ripped out when the time comes with very little work/impact. And as you can see, they're not very significant either (although they will need to grow slightly to be production quality).
2. In the event that the second compilation run errors, the user visible behaviour is completely unchanged.
3. In the event that the second compilation succeeds, the user visible behaviour will be completely unchanged (this isn't entirely true of this PoC, but will be), apart from the fact that auto-completion is cleverer
4. The performance hit is minimal (worst case, it will double the time required to handle `lsp/check`) and zero for programs without errors.
5. The true fix, a proper incremental compiler, is not the work of a moment, and therefore not likely to happen any time soon.

In any case, digging through this has taught me a lot more about how the language server works, and resulted in a couple of long standing language server bugs being fixed in passing, so it's an overall win whether we pursue this or not.

**Comment by magnus-madsen** [comment]
What happens if you write code (like you have with `th`) and you trigger e.g. rename?

File: main/src/ca/uwaterloo/flix/api/lsp/LanguageServer.scala [hunk 1]
```diff
@@ -287,6 +292,36 @@ class LanguageServer(port: Int) extends WebSocketServer(new InetSocketAddress("l
...
+    (prefix ++ (linePrefix + " ??? " + lineRest) ++ rest).mkString
+  }
+
+  private def checkSource(): Validation[TypedAst.Root, CompilationMessage] = {
+    val result = flix.check()
+    result match {
+      case r: Success[_, _] => r
+
+      case Failure(errors) =>
+        val loc = errors.head.loc
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/api/lsp/LanguageServer.scala:309]: Plausibly this should only be done if there is exactly 1 error?

Comment (paulbutcher) [reply main/src/ca/uwaterloo/flix/api/lsp/LanguageServer.scala:309]: I'm not sure I see any benefit in this. In many languages (perhaps not in Flix, I don't have enough experience to be sure) a single typo can result in many subsequent errors. If adding `???` turns (say) 3 errors into zero errors, it's a pretty good guess that that's the case.

And if it turns 3 errors into 2, then we drop the result of the compilation on the floor, so no harm done.

**Comment by magnus-madsen** [comment]
How about errors:

image

I guess the problem is that now ??? has been inserted and the program is correct?

**Comment by paulbutcher** [comment]
Yeah, that's one of the outstanding loose ends (as I said above):

> In the event that the second compilation succeeds, the user visible behaviour will be completely unchanged (this isn't entirely true of this PoC, but will be)

Sorry - I should have made it clearer what those loose ends were.

It's easy to fix: just return the error from the failed compilation from `checkSource`.

**Comment by paulbutcher** [comment]
For completeness, the other loose ends are are around coping with edge cases like a source position that's no longer valid w.r.t. the source in its current form. Plus (obviously) lots more documentation!

**Comment by paulbutcher** [comment]
Have you had any further thoughts on this, @magnus-madsen?

**Comment by magnus-madsen** [comment]
> Have you had any further thoughts on this, @magnus-madsen?

I have not had time  to think it through carefully, but I am very nervous of it leading to a lot of strange behavior (for all the other LSP features). 

E.g. error messages disappearing, renaming working strangely, source locations out of place, all that stuff :/

**Comment by magnus-madsen** [comment]
Perhaps instead of changing LanguageServer directly, we should be doing something just inside `CompletionProvider` (but of course that has some negative architectural consequences.)

**Comment by magnus-madsen** [comment]
For me it is very important that the source code is 100% in sync with the text in the editor.

**Comment by paulbutcher** [comment]
> For me it is very important that the source code is 100% in sync with the text in the editor.

I understand where this comes from. But I guess you're aware that it's already not true? In fact it's almost never true while you're editing the file.

The AST within the language server only ever represents the last successful compilation, which is often quite old while the user is entering text, and can be *very* out of step with the source code if the user is performing major surgery (e.g. copying and pasting chunks of source code).

**Comment by magnus-madsen** [comment]
> > For me it is very important that the source code is 100% in sync with the text in the editor.
> 
> I understand where this comes from. But I guess you're aware that it's already not true? In fact it's almost never true while you're editing the file.
> 
> The AST within the language server only ever represents the last successful compilation, which is often quite old while the user is entering text, and can be _very_ out of step with the source code if the user is performing major surgery (e.g. copying and pasting chunks of source code).

Yes, that is a good point. But, if the program checks (code is green or no errors), then it is in sync.

**Comment by paulbutcher** [comment]
> Yes, that is a good point. But, if the program checks (code is green or no errors), then it is in sync.

Right. And that isn't true for this PoC because I didn't want to invest the effort if it wasn't going to fly. But it will be true once I've tidied up the loose ends.

Perhaps I should go ahead and tidy up the loose ends so you can see it (and see that the source code will never be out of sync if it's green or has no errors).

**Comment by magnus-madsen** [comment]
> > Yes, that is a good point. But, if the program checks (code is green or no errors), then it is in sync.
> 
> Right. And that isn't true for this PoC because I didn't want to invest the effort if it wasn't going to fly. But it will be true once I've tidied up the loose ends.
> 
> Perhaps I should go ahead and tidy up the loose ends so you can see it (and see that the source code will never be out of sync if it's green or has no errors).

If its not too much work you could. Otherwise, just let it simmer a bit in my head. Either I will become comfortable with it or I may come up with a better idea :)

**Comment by paulbutcher** [comment]
It's not a lot of work - I'll do it and you'll be able to allow the full version to simmer instead of this PoC 😉

**Comment by paulbutcher** [comment]
I've checked in changes which:

* Ensure that the error messages seen by the user are always in sync with the current version of the source
* Handle edge cases such as position being out of bounds
* Add comments explaining what's going on 

Note: `checkSource` could be a little nicer if `Validation` implemented `foreach` (allowing `Validation`s to be used within Scala's `for`). But I don't know whether it's worth adding?

**Comment by magnus-madsen** [comment]
Ill give it a spin in the weekend.

**Comment by paulbutcher** [comment]
> Ill give it a spin in the weekend.

Did you have a chance to take another look at this @magnus-madsen?

**Comment by magnus-madsen** [comment]
Another appealing option is the following:

We extend the LanguageServer to track the compilation errors from last compilation. This is a list which is reset when compilation succeeds.

We pass `autoComplete` either the `root` or the `errors`.

if the `errors` contain a `ResolutionError.UndefinedName`:

https://github.com/flix/flix/blob/master/main/src/ca/uwaterloo/flix/language/errors/ResolutionError.scala#L410

(and we are on that line) then we trigger an auto-complete with the local variables current in scope.

One question is how  to get these, but we could extend the error message to have a list of variables that are in scope.

**Comment by magnus-madsen** [comment]
See https://github.com/flix/flix/pull/4159

**Comment by magnus-madsen** [comment]
image

**Comment by magnus-madsen** [comment]
Also works in the middle of an expression, if it parses:

image

**Comment by paulbutcher** [comment]
> I have a proposal for an alternate implementation I would be more comfortable with.

Got you. I'll see about getting this into that form 👍

**Comment by paulbutcher** [comment]
> > I have a proposal for an alternate implementation I would be more comfortable with.
> 
> Got you. I'll see about getting this into that form 👍

Looking at this some more. I can definitely do it, but I don't think it can work in exactly the way you have in mind. Before I forge ahead, I want to make sure that you're OK with this:

The signature of `autoComplete` needs to change to:

```
def autoComplete(uri: String, pos: Position, source: Option[String], current: Boolean, flix: Flix)(implicit index: Index, root: TypedAst.Root): JObject
```

The reasons for this are:

* Even in the event that compilation fails, we still want to pass in the "old" `index` and `root` because these can (and should) still be used for completions.
* We therefore pass in `current` from `LanguageServer` to indicate whether `index` and `root` are in step with the source (and whether `CompletionProvider` should try a second compilation).
* We need to pass in `flix` because we want to use the same `Flix` instance as `LanguageServer` to benefit from incremental compilation.

Sound reasonable?

**Comment by magnus-madsen** [comment]
OK to close this in favor of #4159?

**Comment by paulbutcher** [comment]
> OK to close this in favor of #4159?

Yup 👍

**Pull request closed without merging**